New Features: Quick Flags; Mass Labeler

marcustyphoon opened the pull request:
### Description
<!--
  What is the goal of this pull request?
  How does it achieve that goal?
  Any other context needed to understand the changes?

  Please properly link any issues that this PR aims to resolve:
  https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

This creates a script to quickly toggle the mature content ~~community~~ labels on your ~~original posts~~ posts and reblogs.

It implements a Quick Tags-style dropdown that lets the user label an individual post and a bulk-edit sidebar option and modal form to easily retroactively flag every post with certain tags or the user's entire blog.

Continuation of #771.

- [x] import #839 with slight modifications
- ~~skip attempts to edit legacy posts (https://discord.com/channels/104051306309644288/327155782934200320/1074380420361224272)~~ no longer needed
- [x] use the mass post editor endpoints
- [x] update UI logic for virtual scroller (probably simplify)
- [x] check error handling code
- [x]  #1450

This PR can be split into two (quick reblogs / mass labeler) or three (two stacked PRs on a util implementation PR) on request; it's quite easy for me to do so feel free to do so. While it's not under active review I find it less maintenance to have it as a single PR to reduce work keeping it up to date.

### Testing steps
<!--
  What is the intended behaviour of this pull request?
  How exactly can a maintainer reproduce it?

  Please assume your reviewer will load the addon in a temporary profile.
  Feel free to upload a configuration file if the setup is complex.
-->

Quick Flags: Enable the feature and confirm that...

- Quick Flags buttons are added to editable post/post draft footers with all relevant footer-related account flags
- Quick Flags buttons are **not** added to ask/submission footers with all relevant footer-related account flags
- Quick Flags buttons open UI when clicked that accurately reflects post community label status
- Quick Flags can be used to:
  - add the mature label to an unlabelled post
  - add a category to an unlabelled post, which will also add the mature label
  - remove a category from a labelled post
  - remove the mature label from a labelled post, which will also remove all categories
- After using Quick Flags, Quick Flags buttons open UI when clicked that accurately reflect new post community label status
- After using Quick Flags, an indicator is shown if the new post community label status differs from that shown by the Tumblr UI
- After using Quick Flags, an indicator is **not** shown if the new post community label status **does not** differ from that shown by the Tumblr UI, e.g. if using Quick Flags to set back the original post state
- Indicators persist if posts are scrolled far off the screen and back with endless scrolling enabled
- An error message appears and Quick Flags checkboxes do not become out of sync if the `bulkCommunityLabel` API fetch fails (disconnected network, etc)

Mass Labeler: Enable the feature and confirm that...

- A Mass Labeler sidebar item appears on the "channel" pages' "posts" views
- The Mass Labeler form appears and has the current blog selected when opened
- Mass Labeler adds content labels to posts
- Mass Labeler adds content labels to only posts with the selected tags/in the selected time range
- Mass Labeler adds only the chosen content labels. For example, if the "violence" category is added, at the same time, to a post with the "drug/alcohol" label and a different post with the "sexual themes" label, each post should only have its label and the added label.
- An error message appears if the `bulkCommunityLabel` API fetch fails (disconnected network, etc)

marcustyphoon committed:
import previous work as "quick flags"

marcustyphoon committed:
implement quick tags style dropdown

AprilSylph commented on src/scripts/quick_flags.js:
```diff
@@ -0,0 +1,239 @@
+
+  const {
+    response: {
+      content = {},
```
```suggestion
      content = [],
```

marcustyphoon committed:
fix default content field

Co-authored-by: April Sylph <28949509+AprilSylph@users.noreply.github.com>

marcustyphoon committed:
refinements; disable checkboxes while processing

70m

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
use createEditRequestBody

marcustyphoon committed:
Update quick_flags.json

marcustyphoon committed:
cleanup cleanup everybody do your share

marcustyphoon committed:
fix probable api bug

marcustyphoon committed:
improve UI consistency on error/with multiple browser tabs

marcustyphoon committed:
clean up formatting

marcustyphoon committed:
Implement bulk flagging sidebar menu

marcustyphoon committed:
Update quick_flags.json

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon commented:
I was going to update this to use the mass post editor endpoint (https://www.tumblr.com/changes/715595761003429888/tuesday-april-25th-2023) to perform the edits in batches of 100, but the page actually uses a legacy SVC endpoint at the moment, which we can't realistically make use of.

marcustyphoon committed:
create bulkCommunityLabel helper

AprilSylph commented on src/util/mega_editor.js:
```diff
@@ -56,3 +56,46 @@ export const megaEdit = async function (postIds, options) {
+ * @returns {Promise<Response>} Response from constructed request
+ */
+export const bulkCommunityLabel = async function (blogName, postIds, options) {
+  formKey ??= await fetch('https://www.tumblr.com/neue_web/iframe/new/text').then(response => {
+    if (response.ok) {
+      return response.text();
+    } else {
+      throw Object.assign(new Error(response.status), { response });
+    }
+  }).then(responseText => {
+    const responseDocument = (new DOMParser()).parseFromString(responseText, 'text/html');
+    return responseDocument.getElementById('tumblr_form_key').getAttribute('content');
+  }).catch(console.error);
```
let's abstract this to its own function rather than repeating it

marcustyphoon committed:
Update quick_flags.js

marcustyphoon commented:
The mass post editor continues to overwrite rather than add, unlike what its UI suggests. Using its endpoint correctly would therefore require sorting posts by their current labels, which I actually am happy to code.

This is, on the one hand, a lot of coding effort (sure) and testing/maintenance effort (problematic) for a feature that I would be surprised if many people used.

On the other hand, I think this existing is good for the completeness of the ecosystem; if we really want to encourage people to community label their content comprehensively, better tools to do that seem like good messaging.

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
fix jsdoc

marcustyphoon committed:
extract form key logic

marcustyphoon committed:
use bulkCommunityLabel endpoint

marcustyphoon committed:
Fix edited post state storage for virtual scroller

marcustyphoon committed:
remove unneeded factoring

marcustyphoon committed:
refer to community labels in UI

marcustyphoon committed:
Update quick_flags.json

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
apply #1276

marcustyphoon committed:
Test code (404 every apifetch)

marcustyphoon committed:
Revert "Test code (404 every apifetch)"

This reverts commit 5bb28dc97627f776a18516d0a51b033c574492f1.

marcustyphoon committed:
simplify warning element structure

marcustyphoon committed:
oops

marcustyphoon committed:
use pseudo element for edited warning

marcustyphoon committed:
Revert "use pseudo element for edited warning"

This reverts commit e2ff75aea06d84310811ccb9fb7276e3f06e0a1d.

marcustyphoon committed:
add control button label

AprilSylph reviewed (changes requested):
I think this needs to be split into two features. "Quick Flags" should be the feature with the control buttons on individual posts, while the mass labeller should be, er, probably yeah that. "Mass Labeller".

marcustyphoon committed:
split functionality into two scripts

marcustyphoon committed:
minor refactors

marcustyphoon committed:
implement #1374

I guess this selector should be a util, huh. or actually the whole editable-posts-get-buttons thing should be exported from control_buttons.js

marcustyphoon committed:
simplify form key storage

could instead be top level await, but that would make scripts silently fail to load on any error which I vehemently hate

marcustyphoon committed:
refactor post request

marcustyphoon committed:
fix quick flags ui in error case

marcustyphoon commented:
- [ ] [rename all "community label" ui to "content label"](https://www.tumblr.com/changes/753647297666203648/tuesday-june-18th-2024)

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
fix accent color

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
apply "content label" rename

marcustyphoon committed:
import quick tags button changes

marcustyphoon committed:
always use limit param

Co-Authored-By: April Sylph <28949509+AprilSylph@users.noreply.github.com>

marcustyphoon committed:
use `after` param

Co-Authored-By: April Sylph <28949509+AprilSylph@users.noreply.github.com>

marcustyphoon commented:
- [ ] address #1610

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
display total fetched count without filtering

marcustyphoon committed:
log errors

marcustyphoon committed:
fix merge

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon commented:
- [ ] fix popup state not being set on open (??? how did I not notice this)

marcustyphoon committed:
set initial popup state on open

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
use new control button util

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
update directory structure

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
use shared components in modal dialogue

marcustyphoon committed:
don't unconditionally fetch form key

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon commented:
Oh, hey.

I think I probably finally found the right handler so that we could inject a function to request that a certain post in the timeline be updated from the server (look for `stateNode.refreshPostInTimeline()`, a function with no arguments). Unfortunately, it's on the edit button, which... Tumblr just removed from the post body. Presumably we could click the meatballs button, then use the edit button in the meatballs menu to request a refresh, but that's pretty absurd.

Possibly more plausible, but considerably more involved, is obtaining the `onEditTimelineObject` function that it calls by grabbing it from... uh, some withContext HOC? and figuring out how to tell it to make the correct change.

Both of these are, obviously, way more than we generally do with injected functions, which means they're pretty fragile to Tumblr code changes and might be difficult or impossible to fix if they did break, which seem like reasonable reasons to continue favoring less elegant functionality that can be done without injecting, like the warning element thingy currently implemented here.

Somewhat related: #1901's "let's put back the edit and delete buttons" thing; I think in that case we'll need to just open the meatballs menu and click it, though (delete could maybe be done by simulating the `d` hotkey, but there's no edit hotkey, so).

marcustyphoon commented:
Hey, want to see some shit that absolutely should not work but does? https://github.com/AprilSylph/XKit-Rewritten/compare/master...AprilSylph:XKit-Rewritten:marcustyphoon/quick-tags-really-refresh

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
quick flags: use new dom utils

marcustyphoon committed:
mass labeler: use new dom utils

marcustyphoon committed:
update warning element for new footer

marcustyphoon committed:
update css for `appendWithoutOverflow` util change

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
remove jsdoc hyphens

marcustyphoon committed:
use ts jsdoc

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
format

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
Merge remote-tracking branch 'upstream/master' into community-labeler-3

marcustyphoon committed:
merge maintenance: text format util

marcustyphoon committed:
remove outdated state warning element

marcustyphoon committed:
use updatePostOnPage util

marcustyphoon committed:
updatePostOnPage util: allow deletion

marcustyphoon committed:
fix outdated community label state

marcustyphoon committed:
Merge branch 'master' into community-labeler-3

marcustyphoon committed:
make post update conditional on edit success

marcustyphoon committed:
clarify comment

marcustyphoon committed:
Merge branch 'master' into community-labeler-3
